man/man8/smtpd.8
.TH "SMTPD" 8 "21 May 85"
.SH NAME
smtpd, smtpsrvr \- SMTP inbound channel for MMDF
.SH SYNOPSIS
smtpd [\-d] [\-f] [\-n maxconnects] smtpsrvr channels
.br
smtpd [\-i] [\-f] smtpsrvr channels
.sp
.br
smtpsrvr remotehost localhost [ channels ]
.br
rsmtpsrvr remotehost localhost [ channels ]  (finicky)
.SH DESCRIPTION
The SMTP inbound channel is part of the MMDF mail system.
It consists of two programs,
an SMTP server (\fIsmptsrvr\fR) and a connection server (\fIsmtpd\fR).
\fISmptsrvr\fR is designed to be called by a networking daemon
that has determined the identity of the caller and set up the network
connection on the server's standard input and output.
\fISmtpsrvr\fR interfaces between the network and the mail system.
It calls \fIsubmit\fR to pass mail into the system.
.PP
The
.I smtpd
program comes in several flavors depending on your operating system.
In some cases, the burden of limiting and waiting for inbound connections
is borne by a standard "daemon of daemons" (options \-i).  In other cases, 
.I smtpd 
performs these operations.  In any case, 
.I smtpd
is the link between the network connection and the
.I smtpsrvr 
program.
.PP
In the synopsis above, the first version of 
.I smtpd
is a standalone network daemons that accept incoming
connections and fork off copies of the smptsrvr to process
them. After preparing itself for accepting connection
.I smtpd 
gives up any root-permissions and set its effectiv-userid to the owner
of the \fIsubmit\fR.
The last version is called by the standard 
``daemon of daemons'' with one connection and fork the smtpsrvr to
process it.
The sun version calls the standard 
.I smtpsrvr
program with channels set to ``smtp''.
.PP
The \-i option tells smtpd that it is started by the super-daemon 'inetd'.
The \-d option enables debugging prints on the stderr output.
The \-f option (the \fBf\fRiniky option) causes 
incoming connections from unknown
hosts to be rejected.
.I Smtpd
notifies the \fIsmtpsrvr\fR program to use the finicky option
by calling it with argv[0] set to `rsmtpsrvr'.
The maxconnects argument sets the maximum number
of simultaneous connections to support.  This value defaults to 4 on 4.2
systems.
.PP
The smtpsrvr argument is the path to the \fIsmtpsrvr\fR
program (relative or absolute).
Generally this is something like /usr/mmdf/chans/smtpsrvr.
.PP
The channels argument is a comma-separated list of channels.  When a connection
is established, 
the source host is looked up in the host tables for these channels.
If the source host is found, the 
corresponding channel name is 
given to \fIsubmit\fR as the channel it should 
consider the mail to have come from for authorization purposes.
In the finicky mode,
if the source host is not found, \fIsmtpsrvr\fR will
reject the incoming mail.  In the normal (non-finicky)
mode, if the source host is not found, \fIsmtpsrvr\fR will assume the
mail came in on the last channel in the channels list.
.br 
The channels argument may be ommited when specified in the \fIMMDF
tailor\fR file (see mmdftailor(5) parameter MVALIDCHN ).
.PP
If your smtp channel table is a nameserver-type channel table and you
are running in the finicky mode, then
you should include the "local" channel in the list of acceptable
channels.  If you do not include "local" in this case, you will not
be allowed to connect to your own server and submit mail because the
nameserver channel will not let you look up your own address.  (RFC 974
requires this to prevent mail loops).
.PP
Since the \fIsmtpsrvr\fR program sets the trustme flag when calling
\fIsubmit\fR,
it must be called by a process whose UID is that
of a trusted mail submitter (generally only root and mmdf).
.SH SEE ALSO
deliver(8), submit(8), smtp(8), RFC821


Changes
Changes beta 3
	* Small bug in ap_lex.c hip@a.cs.okstate.edu
	* jromine@ics.uci.edu some changes for Solaris 2.5, Sun Pro
	  4.2 Compiler
	* linux 2.2.7, glibc2.1, egcs 2.91 (redhat 6.0)
	* sco (first attempt)
	* jromine@ics.uci.edu, boundary fix smtpsrvr
	* Small bug in submit/lnk_getaddr.c krueger@mathematik.uni-kl.de

Wed Dec 31 13:14:34 MET 1997	Kai Krueger <krueger@mathematik.uni-kl.de>

	* create file h/config.h.in and moved every define to it.

	* DOASSIGN renamed to HAVE_ASSIGN. Autodetection by configure

	* SECURETTY renamed to HAVE_SECURETTY, Autodetection by configure

	* NODIAL renamed to HAVE_DIAL. default is 'no'
	  may be enanble by configure --enable-dial
	  if compiling phone/smphone channel it wil be turned on by configure.

	* NAMESERVER renamed to HAVE_NAMESERVER. Autodetection by configure.
	  May be overwriten by --disable-nameserver
          - tb_ns.4.2.c and tb_ns.fake.c merged into one file tb_ns.c and
	  seperated by #ifdef HAVE_NAMESERVER...
	  - removed TB_NS in Makefile.com	  

	* HAVE_TCP_WRAPPER renamed to HAVE_LIBWRAP. Autodetection by configure.
	
	* <time.h> included in config.h. It's location is determined by
	  configure

	* determination of <dir.h>, <dirent.h>, a.s.o is done by configure
	  definition an include has moved to config.h

	* determination of index/rindex done by configure and moved to config.h

	* definition of sigtype renamed to RETSIGTYPE. Detection done by
	  configure.

	* return-type of sprintf determined by configure. Define BSD_SPRINTF
	  if char * is returned.

	* uip/Makefile, src/Makefile, lib/Makefile Makefile.com and Makefile
	  are generated by configure.

	* Value of CH_TB in Makefile.com set by configure. If you want
	  sequential tables run configure with the argument --disable-dbm

	* Added define HAVE_SGTTY_H for the existens of <sgtty.h>

	* Added check for <varargs.h> as define HAVE_VARARGS_H.
	  Added check for variable number of arguments if <varargs.h> don't
	  exists. If Compiler don't supports variable number of arguments
	  NO_VARARGS is defined. (NO_VARARGS should be renamed)
	  configure ist checking for both of it. But if <varagrs.h> isn't
	  found lib/util/multcat.c and lib/util/multcpy.c could't be compiled.
	  
	* test configure on machine that dosn't support multiple arguments.
	  (check of NO_VARARGS)

	* test for dup2() and fcntl(fd1, F_DUPFD, fd2) done.
	  NODUP2 replaced by HAVE_DUP2 and
	  NOFCTNL replaced by HAVE_FCNTL_F_DUPFD


src/smtp/smtpd.c
/*
 *			S M T P D . C
 *
 * Server SMTP daemon, for 4.1a/4.2 BSD by DPK at BRL, 1 Jan 82 (Sigh...)
 *
 *		R E V I S I O N   H I S T O R Y
 *
 *	02/02/83  MJM   Modified to be more robust.
 *
 *	03/18/83  DPK   Modified for 4.1c.
 *
 *	??		Modified for 4.2
 */

#include "util.h"
#include <signal.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>
#include <sys/wait.h>
#include <sys/ioctl.h>
#include <sys/stat.h>

#include "ns.h"
/*#ifndef LINESIZE
#  define LINESIZE 1024
#endif*/

extern	int	errno;

struct	sockaddr_in	addr;

char	*Smtpserver = 0;	/* Actual smtp server process, full path */
char	*Channel = 0;		/* Logical channel mail will arrive on */
int	Maxconnections = 4;	/* Maximum simultaneous connections */
int	numconnections = 0;	/* Number of currently active connections */
int	debug = 0;		/* If nonzero, give verbose output on stderr */
#define	logx	if (debug) log
char	errbuf[BUFSIZ];		/* Logging will be line buffered */
char	programid[40];		/* identification string for log entries */
int	stricked;		/* force full validity of source address */
int     started_by_inetd;

int     effecid,                  /* system number of pgm/file's owner  */
	callerid;                 /* who invoked me?                    */

extern char *inet_ntoa();
LOCFUN mn_mmdf();
RETSIGTYPE sig17();

main (argc, argv)
int argc;
char **argv;
{
	register	int		skt;
			struct	servent	*sp;
			struct	hostent	*hp;
			char		thishost[64];
			char		workarea[32];
			int		i, pid;

	setbuf(stderr, errbuf);
	sprintf(programid, "smtpd(%5.5d): ", getpid());
	getwho (&callerid, &effecid); /* who am I and who is running me?    */
	mmdf_init (argv[0]);
	gethostname(thishost, sizeof(thishost));	/* current hostname */
	if ((sp = getservbyname("smtp", "tcp")) == NULL) {
		fprintf(stderr, "Cannot find service smtp/tcp\n");
		exit(-1);
	}
#ifdef HAVE_NAMESERVER
	/* don't wait forever! */
	ns_settimeo(NS_NETTIME);
#endif /* HAVE_NAMESERVER */

	/*
	 * try to get full name for thishost
	 */
	hp = gethostbyname(thishost);
	if (hp != NULL)	strcpy(thishost, hp->h_name);

	/* Parse args; args will override configuration file... */
	arginit(argc, argv);

	if(started_by_inetd) {
	  char *rmthost;
	  struct sockaddr_in rmtaddr;
	  int     len_rmtaddr = sizeof rmtaddr;
	  int     on = 1;
	
	  mn_mmdf();           /* set up effective and group id's properly */
	  if (getpeername (0, (struct sockaddr *)&rmtaddr, &len_rmtaddr) < 0)
	    bomb( "getpeername failed (errno %d)",errno);

	  (void) setsockopt (0, SOL_SOCKET, SO_KEEPALIVE, &on, sizeof on);
	  hp = gethostbyaddr ( (char *)&rmtaddr.sin_addr,
			       sizeof(rmtaddr.sin_addr), AF_INET );
	  if ((hp == NULL) || !isstr(hp->h_name)) {
	    strcpy(workarea, (char *)inet_ntoa(rmtaddr.sin_addr));
	    rmthost = workarea;
	  } else
	    rmthost = hp->h_name;
	  
	  execl (Smtpserver, stricked ? "rsmtpsrvr" : "smtpsrvr",
		 rmthost, thishost, Channel, (char *)0);
	  bomb( "server exec error (%d)", errno);
	}
	/* now smtpd is running standalone */

	addr.sin_family = AF_INET;
	addr.sin_addr.s_addr = INADDR_ANY;
	addr.sin_port = (short)sp->s_port;

#ifdef never
	if ((i = open("/dev/tty", 2)) >= 0) {
		(void) ioctl(i, TIOCNOTTY, (char *)0);
		(void) close(i);
	}
#endif

	/*
	 * Create a socket to accept a SMTP connection on.
	 * The doloop is due to some sort of IPC bug in 4.1a.
	 */
	logx("opening socket...");
	i = 60;
	do {
		skt = socket(AF_INET, SOCK_STREAM, 0);
		logx("socket returned %d", skt);
		if (skt < 0)
			sleep(5);
	} while (skt < 0 && i-- > 0);
	if (skt < 0) {
		logx("can't open socket (errno %d)", errno);
		exit(99);
	}
	logx("socket open on %d", skt);

	if (bind (skt, (char *)&addr, sizeof addr) < 0) {
		logx("can't bind socket (errno %d)", errno);
		exit(98);
	}
	listen (skt, Maxconnections+1);
	logx("old with uid=%d, euid=%d", getuid(), geteuid());
	logx("old with gid=%d, egid=%d", getgid(), getegid());
	mn_mmdf();           /* set up effective and group id's properly */

	logx("running with uid=%d, euid=%d", getuid(), geteuid());
	logx("running with gid=%d, egid=%d", getgid(), getegid());
	signal( SIGCHLD, sig17);

	while (1) {
		extern	char			*inet_ntoa();
			struct	sockaddr_in	rmtaddr;
			int			len_rmtaddr = sizeof rmtaddr;
			int			status;
			int			tmpskt;

		/*
		 * Accept a connection.
		 */
		if ((tmpskt = accept(skt, &rmtaddr, &len_rmtaddr)) < 0) {
			logx("accept error (%d)", errno);
			sleep(1);
			continue;
		}

		/* We have a valid open connection, start a server... */
		if ((pid = fork()) < 0) {
			logx("could not fork (%d)", errno);
			close(tmpskt);
		} else if (pid == 0) {
			/*
			 * Child
			 */
			char	*rmt;

			hp = gethostbyaddr(&rmtaddr.sin_addr,
					sizeof(rmtaddr.sin_addr), AF_INET);
			if ((hp == NULL) || !isstr(hp->h_name)) {
				strcpy(workarea, inet_ntoa(rmtaddr.sin_addr));
				rmt = workarea;
			}
			else
				rmt = hp->h_name;

			logx("(%d) %s started", numconnections, rmt);
			dup2(tmpskt, 0);
			dup2(tmpskt, 1);
			if (tmpskt > 1)
				close(tmpskt);
			execl(Smtpserver, stricked? "rsmtpsrvr" : "smtpsrvr",
					rmt, thishost, Channel, (char *)0);
			logx("server exec error (%d)", errno);
			exit(99);
		}

		/*
		 * Parent
		 */
		close(tmpskt);
		numconnections++;

		/*
		 * This code collects ZOMBIES and implements load
		 * limiting by staying in the do loop while the
		 * Maxconnections active.
		 */
#if HAVE_WAIT3
		while (wait3 (&status, numconnections < Maxconnections
			      ? WNOHANG : 0, 0) > 0)
#else /* HAVE_WAIT3 */
		  while (waitpid (-1, &status, numconnections < Maxconnections
				  ? WNOHANG : 0) > 0)
#endif /* HAVE_WAIT3 */
		    numconnections--;
	}
}

arginit(argc, argv)
int argc;
char **argv;
{
	register	int	ch;
	extern		char	*optarg;
	extern		int	optind;

	while ((ch = getopt(argc, argv, "difn:")) != EOF) {
		switch (ch) {
		case 'd':
			debug++;
			break;

		case 'i':	/* force correctness of addresses */
			started_by_inetd++;
			break;

		case 'f':	/* force correctness of addresses */
			stricked++;
			break;

		case 'n':
			Maxconnections = atoi(optarg);
			if (Maxconnections <= 0) {
				logx("Bad number of connections '%s'", optarg);
				exit(99);
			}
			logx("Maxconnection now %d", Maxconnections);
			break;

		default:
			log("Usage: %s [-dif] [-n #connections] smtpserver channels", argv[0]);
			exit(1);
		}
	}

	if (optind == argc) {
		log("Usage: %s [-dif] [-n #connections] smtpserver channels",
			argv[0]);
		exit(1);
	}

	if (optind < argc) {
		if (access(argv[optind], 01) < 0) {	/* execute privs? */
			logx("Cannot access server program '%s'", argv[optind]);
			exit(99);
		}
		Smtpserver = argv[optind];
		logx("server is '%s'", Smtpserver);
	} else {
		logx("Smtpserver program not specified!");
		exit(99);
	}

	if (++optind < argc) {
		Channel = argv[optind];
		logx("channel is '%s'", Channel);
	} else {
		logx("Channel not specified!");
		exit(99);
	}
}

/*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*/
RETSIGTYPE sig17()
{
  int	status, pid;
  signal( SIGCHLD, sig17);
#if HAVE_WAIT3
  pid = wait3 (&status, WNOHANG, 0);
#else /* HAVE_WAIT3 */
  pid = waitpid (-1, &status,  WNOHANG);
#endif /* HAVE_WAIT3 */
  if(pid>0) numconnections--;
}

/*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*/
/* VARARGS */
log(fmt, a, b, c, d)
char *fmt;
int a, b, c, d;
{
	fputs(programid, stderr);
	fprintf(stderr, fmt, a, b, c, d);
	fputc('\n', stderr);
	fflush(stderr);
}

/*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*/
bomb( fmt, a, b, c, d )
char *fmt;
{
        fputs( "451 ", stdout );
        fprintf( stdout, fmt, a, b, c, d );
        fputs( "\r\n", stdout );
        fflush( stdout );
        exit (99);
}

/*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*/
LOCFUN
        mn_mmdf ()		  /* setuid to mmdf: bypass being root  */
{				  /* get sys id for mmdf; setuid to it  */
    extern char *pathsubmit;     /* submit command file name           */
    extern char *cmddfldir;      /* directory w/mmdf commands          */
    char    temppath[LINESIZE];
    struct stat    statbuf;

#ifdef DEBUG
    logx ("mn_mmdf(); effec==%d", effecid );
#endif

/*  the following is a little strange, doing a stat on the object
 *  file, because setuid-on-execute does not work when the caller
 *  is root, as will happen when this is started by the rc file.
 *  hence, the effective id, from a getuid, will show root & not mmdf.
 *
 *  the goal is to have this process be name-independent of the caller,
 *  so that returned mail comes from mmdf and not the invoker.
 *
 *  in pickup mode, the id of the caller has to be retained, since
 *  pobox channels use that to determine access rights to mail.
 *
 *  All sets gid to mmdf's gid  --  <DPK@BRL>
 */

    if (effecid == 0)
    {
	getfpath (pathsubmit, cmddfldir, temppath);

	if (stat (temppath, &statbuf) < OK)
	    bomb ("Unable to stat %s", temppath);
				  /* use "submit" to get mmdf id        */

	if (setgid (statbuf.st_gid) == NOTOK)
	    bomb ("Can't setgid to mmdf (%d)", statbuf.st_gid);
	if (setuid (statbuf.st_uid) == NOTOK)
	    bomb ("Can't setuid to mmdf (%d)", statbuf.st_uid);

	effecid = statbuf.st_uid; /* mostly needed for return mail      */
    }
}
/*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*/
